### ISSUE TITLE
Regression: unable to define bindings via `bindcode` without using a modifier (Mod2, Mod4)

### ISSUE THREAD

**Issue by maximbaz** [i3/i3#2609]
Output of `i3 --moreversion 2>&- || i3 --version`:

```
λ:  i3 --moreversion 2>&- || i3 --version
Binary i3 version:  4.13 (2016-11-08) © 2009 Michael Stapelberg and contributors
Running i3 version: 4.13 (2016-11-08) (pid 2030)o abort…)
Loaded i3 config: /home/maximbaz/.i3/config (Last modified: Sun 18 Dec 2016 11:24:06 PM CET, 76 seconds ago)

The i3 binary you just called: /usr/bin/i3
The i3 binary you are running: i3
```

URL to a logfile as per http://i3wm.org/docs/debugging.html:

http://logs.i3wm.org/logs/5636647567753216.bz2

**What I did:**

Use the following `.i3/config`:

```
set $mod Mod4

set $mode_resize Resize window
mode "$mode_resize" {
  bindcode 43            resize grow   left  10 px or 10 ppt
  bindcode 46            resize shrink left  10 px or 10 ppt

  # bindsym h            resize grow   left  10 px or 10 ppt
  # bindsym l            resize shrink left  10 px or 10 ppt

  bindsym Return mode "default"
  bindsym Escape mode "default"
}
bindcode $mod+27 mode "$mode_resize"
```

1. Hit `Mod4+R` - make sure i3wm processes my binding and enters the "Resize" mode.
2. Hit `L` or `H` keys in an attempt to resize the window.

**What I saw:**

1. `Mod4+R` worked fine, I entered the "Resize" mode, so normal `bindcode` works.
2. The `L` or `H` keys do not resize the window, simply nothing happens as I press them, so `bindcode` does not work within a `mode`.

**What I expected instead:**

The `L` and `H` keys work when the configuration consists of `bindcode` definitions.

**Extra info:**

Comment out `bindcode` definitions inside the `mode` and uncomment the `bindsym`.
Now `L` and `H` keys do resize the window as expected, so `bindsym` works within a `mode`.

**This is a regression from 4.12.1**. I do not change the config, install `4.12.1`, restart `i3` and the keys `L` and `H` work as expected with the `bindcode` definition.

```
λ:  i3 --moreversion 2>&- || i3 --version
Binary i3 version:  4.12 (2016-03-06, branch "4.12") © 2009 Michael Stapelberg and contributors
Running i3 version: 4.12 (2016-03-06, branch "4.12") (pid 2030)
Loaded i3 config: /home/maximbaz/.i3/config (Last modified: Sun 18 Dec 2016 11:24:06 PM CET, 998 seconds ago)

The i3 binary you just called: /usr/bin/i3
The i3 binary you are running: i3
```

**Comment by Airblader** [comment]
FYI: I can't reproduce this. `bindcode` works for me, even within in a binding mode.

**Comment by maximbaz** [comment]
Even with my exact config file? Interesting, I can consistently reproduce it by installing `4.13-1` and `4.12-1` back and forth, and restarting the i3.

**Comment by Airblader** [comment]
The log file says your resize mode actually contained all of these:

````
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 181): # Resize window (r)
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 182): set Resize window Resize window
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 183): mode "Resize window" {
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 184):   # (h)
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 185):   bindcode 43            resize grow   left  10 px or 10 ppt
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 186):   bindcode Shift+43      resize shrink right 10 px or 10 ppt
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 187):   bindcode Ctrl+43       resize grow   left  1  px or 1  ppt
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 188):   bindcode Ctrl+Shift+43 resize shrink right 1  px or 1  ppt
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 189):   # (j)
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 190):   bindcode Shift+44      resize grow   down  10 px or 10 ppt
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 191):   bindcode 44            resize shrink up    10 px or 10 ppt
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 192):   bindcode Ctrl+Shift+44 resize grow   down  1  px or 1  ppt
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 193):   bindcode Ctrl+44       resize shrink up    1  px or 1  ppt
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 194):   # (k)
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 195):   bindcode Shift+45      resize shrink down  10 px or 10 ppt
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 196):   bindcode 45            resize grow   up    10 px or 10 ppt
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 197):   bindcode Ctrl+Shift+45 resize shrink down  1  px or 1  ppt
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 198):   bindcode Ctrl+45       resize grow   up    1  px or 1  ppt
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 199):   # (l)
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 200):   bindcode Shift+46      resize grow   right 10 px or 10 ppt
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 201):   bindcode 46            resize shrink left  10 px or 10 ppt
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 202):   bindcode Ctrl+Shift+46 resize grow   right 1  px or 1  ppt
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 203):   bindcode Ctrl+46       resize shrink left  1  px or 1  ppt
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 204): 
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 205):   # back to normal: Enter or Escape
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 206):   bindsym Return mode "default"
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 207):   bindsym Escape mode "default"
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 208): }
12/18/2016 11:17:49 PM - config_parser.c:parse_config:267 - CONFIG(line 209): bindcode Mod4+27 mode "Resize window"
````

I assume the bindings in your System binding mode also don't work?

**Comment by maximbaz** [comment]
I came up with this minimal config file after I uploaded the log. 
But I've just re-uploaded the log with that specific configuration that I posted.

http://logs.i3wm.org/logs/5635663080718336.bz2

**Comment by maximbaz** [comment]
Just for completeness, this is a log for `4.12-1`, also with that specific configuration and where the bindings actually do resize the window.

http://logs.i3wm.org/logs/5762244625301504.bz2

**Comment by Airblader** [comment]
Although I currently don't have time to look into it, just wanted to say thanks for providing all the log files. I'm sure they'll be helpful once somebody works on this. :-)

**Comment by maximbaz** [comment]
Sounds great :+1: 

Once somebody gets to this, please do not hesitate to ping me, I'm happy to provide some extra information or to test possible solutions.

**Comment by Airblader** [comment]
@maximbaz Somebody on reddit reported a similar issue where some bindcode commands didn't work, though it had nothing to do with binding modes. They reported that while 4.13 has the issue, the current `next` branch doesn't.

I'm not aware of any change we pushed since 4.13 that would fix this, but if you can give it a shot and – if it works now – give it a git bisect, that'd be great.

**Comment by maximbaz** [comment]
That is a good suggestion! I've just tried the `next` branch, unfortunately the issue is still there.
However I'll try now to bisect between 4.12 and 4.13, and hopefully I'll find that commit.

**Comment by maximbaz** [comment]
@Airblader according to `git bisect`, d77d40173aae250d2e5ba86a8bce7a6e857cbc67 is the first bad commit.

**Comment by maximbaz** [comment]
A friendly ping to @stapelberg: in case you have any ideas what might have gone wrong, please share :)

**Comment by maximbaz** [comment]
Update: I stumbled upon #2558 and #2559, and decided to try to add `Mod2+` to every `bindcode` within a mode. Guess what, the bindings stopped working in 4.12, but they started to work in 4.13! 

Adding `Mod2+` to normal bindings (outside of `mode`) is also interesting. In 4.12 they stop working, but in 4.13 the bindings work both with and without this `Mod2+` prefix.

Interestingly enough, I don't even have a `NumLock` key on my laptop (that is the meaning of `Mod2`, right?). Thus I see no reason why this prefix is required, especially only within the `mode`. 

I'm glad that workaround exists, but I think this issue still needs to be looked at some point.

**Comment by Airblader** [comment]
Yeah, it's definitely weird that it requires Mod2 in a binding mode but not outside of it (have you tried the exact same keycode outsode of a binding mode?). This is really more @stapelberg's territory than mine, though. :-)

**Comment by maximbaz** [comment]
Your suggestion brought me to a very interesting discovery! Apparently all the `bindcode` work in "default" mode only because by default all my key bindings have the prefix `$mod+`, where `$mod = Mod4` :confused:

Therefore, this config is the proper minimal config to reproduce the issue:

```
set $mod Mod4
bindcode $mod+Shift+27 restart

bindcode 24        exec notify-send "(1) q in normal mode"
bindcode $mod+24   exec notify-send "(2) q with Mod4 in normal mode"
bindcode Mod2+24   exec notify-send "(3) q with Mod2 in normal mode"

set $mode_test Test
mode "$mode_test" {
  bindcode 24      exec notify-send "(4) q in test mode"
  bindcode $mod+24 exec notify-send "(5) q with Mod4 in test mode"
  bindcode Mod2+24 exec notify-send "(6) q with Mod2 in test mode"
  bindsym Escape mode "default"
}
bindcode $mod+27 mode "$mode_test"
```

* With 4.12 ([log](http://logs.i3wm.org/logs/5169744760012800.bz2)):
  * Press `Q` => Triggers (1)
  * Press `Win+Q` => Triggers (2)
  * Enter Test mode
  * Press `Q` => Triggers (4)
  * Press `Win+Q` => Triggers (5)
  * Remember I don't have NumLock to even trigger bindings (3) and (6).
* With 4.13 ([log](http://logs.i3wm.org/logs/5661601461960704.bz2)):
  * **???** Press `Q` => Triggers (3)
  * Press `Win+Q` => Triggers (2)
  * Enter Test mode
  * **???** Press `Q` => Triggers (6)
  * Press `Win+Q` => Triggers (5)

The items marked with **???** are the ones that require further attention.

**Comment by maximbaz** [comment]
I renamed the issue since it's clear that the issue is not with modes, but with defining a binding without specifying any of the modifiers.

**Comment by Airblader** [comment]
Ah, we are finally closing in on this, I think. And in fact I very much suspect that this is behaving as expected (we did anticipate breakage with the 4.13 config, unfortunately). You may not have a NumLock key on your keyboard, but that doesn't mean there isn't a NumLock flag somewhere. :-)

You can try installing `numlockx` and using that to turn off NumLock via `numlockx off`. I can actually reproduce this with

    bindcode 27 exec …

with NumLock on this won't work, but with NumLock off it does work.

So in conclusion I think this is basically #2558 and might also be related to #2559

**Comment by maximbaz** [comment]
I see... I understand that sometimes breaking changes are needed, that is totally fine, but in this specific case I'd like to clarify a few questions:
* Why are the bindings for letters like `q` different with and without `NumLock`? I understand why it is useful to recognize the meaning of `KP_` keys, they do have two meanings with and without `NumLock` pressed. But letters like `q`?
* What is the new recommended way to define bindings? Most of my bindings are not dependent on `NumLock` state. Should I duplicate each of them with and without `Mod2`, just in case? 
* Why the cases (2) and (5) even work today? Shouldn't they also contain a `Mod2` prefix? Since the `q` is prohibited in favor of `Mod2+q`, shouldn't `Win+q` be also prohibited in favor of `Win+Mod2+q`?

**Comment by Airblader** [comment]
> Why are the bindings for letters like q different with and without NumLock?

See #2259, that's exactly the discussion there.

> Most of my bindings are not dependent on NumLock state. Should I duplicate each of them with and without Mod2, just in case?

If you don't ever use NumLock you don't need to duplicate the bindings. Just make sure your NumLock is set to a constant value (whether it be on or off) and configure bindings accordingly.

> Why the cases (2) and (5) even work today? Shouldn't they also contain a Mod2 prefix?

It's all about what happens in the end. I assume that `Win` temporarily clears the NumLock state. But this isn't really my forte, so don't take my word for it. :-)

**Comment by maximbaz** [comment]
I posted the last config and logs in #2559, let us switch to that issue then.
I will close this issue as duplicate, since there's nothing else to investigate apart from the #2559 itself.
